### ISSUE TITLE
Null-dereference READ (9283)

### ISSUE THREAD

**Issue by dlemstra** [strukturag/libheif#48]
The @ImageMagick project is using https://github.com/google/oss-fuzz to find bugs in our own library and in libraries that we use. The fuzzer found an issue and we think this is an issue that should be resolved in the library that we use. This issue is posted under the url https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=9283 that is not publicly visible yet but added as a link for future reference. Below are the details of the issue that can be reproduced using the following technique: https://github.com/google/oss-fuzz/blob/master/docs/reproducing.md

Stacktrace:
```
SCARINESS: 10 (null-deref)
--
  | #0 0x7fb55a54dba5 in memcpy-avx-unaligned.S:180 /build/glibc-Cl5G7W/glibc-2.23/sysdeps/x86_64/multiarch/memcpy-avx-unaligned.S:180
  | #1 0x4eba2a in __asan_memcpy _asan_rtl_
  | #2 0xdaadca in heif::HeifContext::decode_and_paste_tile_image(unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int) const libheif/libheif/heif_context.cc:1148:7
  | #3 0xdbe62a in __invoke<heif::Error (heif::HeifContext::*)(unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int) const, const heif::HeifContext *, unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int, void> /usr/local/include/c++/v1/type_traits:4286:1
  | #4 0xdbe62a in heif::Error std::__1::__async_func<heif::Error (heif::HeifContext::*)(unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int) const, heif::HeifContext const*, unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int>::__execute<1ul, 2ul, 3ul, 4ul, 5ul>(std::__1::__tuple_indices<1ul, 2ul, 3ul, 4ul, 5ul>) /usr/local/include/c++/v1/future:2330
  | #5 0xdbe3a8 in std::__1::__async_func<heif::Error (heif::HeifContext::*)(unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int) const, heif::HeifContext const*, unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int>::operator()() /usr/local/include/c++/v1/future:2323:16
  | #6 0xdbdb8a in std::__1::__async_assoc_state<heif::Error, std::__1::__async_func<heif::Error (heif::HeifContext::*)(unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int) const, heif::HeifContext const*, unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int> >::__execute() /usr/local/include/c++/v1/future:985:25
  | #7 0xdbe884 in __invoke<void (std::__1::__async_assoc_state<heif::Error, std::__1::__async_func<heif::Error (heif::HeifContext::*)(unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int) const, const heif::HeifContext *, unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int> >::*)(), std::__1::__async_assoc_state<heif::Error, std::__1::__async_func<heif::Error (heif::HeifContext::*)(unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int) const, const heif::HeifContext *, unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int> > *, void> /usr/local/include/c++/v1/type_traits:4286:1
  | #8 0xdbe884 in __thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (std::__1::__async_assoc_state<heif::Error, std::__1::__async_func<heif::Error (heif::HeifContext::*)(unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int) const, const heif::HeifContext *, unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int> >::*)(), std::__1::__async_assoc_state<heif::Error, std::__1::__async_func<heif::Error (heif::HeifContext::*)(unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int) const, const heif::HeifContext *, unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int> > *, 2> /usr/local/include/c++/v1/thread:342
  | #9 0xdbe884 in void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (std::__1::__async_assoc_state<heif::Error, std::__1::__async_func<heif::Error (heif::HeifContext::*)(unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int) const, heif::HeifContext const*, unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int> >::*)(), std::__1::__async_assoc_state<heif::Error, std::__1::__async_func<heif::Error (heif::HeifContext::*)(unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int) const, heif::HeifContext const*, unsigned int, std::__1::shared_ptr<heif::HeifPixelImage>, int, int> >*> >(void*) /usr/local/include/c++/v1/thread:352
  | #10 0x7fb55adf36b9 in start_thread
  | #11 0x7fb55a50741c in clone /build/glibc-Cl5G7W/glibc-2.23/sysdeps/unix/sysv/linux/x86_64/clone.S:109
```

Links: https://github.com/strukturag/libheif/blob/01fe0a821d5feefb5e8cdefbf2310a07b25b04a8/libheif/heif_context.cc#L1148

Commit: 01fe0a821d5feefb5e8cdefbf2310a07b25b04a8
Testcase file: [clusterfuzz-testcase-minimized-encoder_heic_fuzzer-4878614100180992.zip](https://github.com/strukturag/libheif/files/2194785/clusterfuzz-testcase-minimized-encoder_heic_fuzzer-4878614100180992.zip)

**Comment by farindk** [comment]
already fixed in master branch (post v1.3.2)

**Comment by fancycode** [comment]
I can reproduce that with the provided testfile, `tile_data` is `NULL` in https://github.com/strukturag/libheif/blob/01fe0a821d5feefb5e8cdefbf2310a07b25b04a8/libheif/heif_context.cc#L1125

This happens with latest master (https://github.com/strukturag/libheif/commit/01fe0a821d5feefb5e8cdefbf2310a07b25b04a8) as described above.

**Comment by farindk** [comment]
fixed in 317302bb4
